File: 김금환/해시/폰켓몬.cpp
#include <vector>
#include <set>

using namespace std;

set<int> types;

int solution(vector<int> nums)
{
    // 종류 수 세기
    for (int num : nums) {
        types.insert(num);
    }
    
    // 종류 수가 선택할 수 있는 폰켓몬의 수보다 크거나 같으면 선택할 수 있는 폰켓몬의 수가 답
    // 종류 수가 선택할 수 있는 폰켓몬의 수보다 작으면 종류 수가 답
    int answer = types.size() >= nums.size() / 2 ? nums.size() / 2 : types.size();
    return answer;
}

File: 이예진/해시/폰켓몬.py
# 방법 1 set 사용
def solution(nums):
    n = len(nums)//2
    nums_s= len(set(nums))
    if n < nums_s:
        return n
    return nums_s

# 방법 2 dictionary 사용
def solution(nums):
    answer = 0
    d = {}
    
    n = len(nums)//2
    
    for i in nums:
        if i not in d:
            d[i] = 1
        else:
            d[i] += 1
    
    if len(d) >= n:
        answer = n
    else:
        answer = len(d)
        
    return answer